Add a button to change harvest pickleader & a dropdown to change status

mina-andajos opened the pull request:
<!-- 
Thanks for your contribution!
Please fill out the necessary sections, and delete unused ones.
-->
Fixes #199
----------
## Type of change:
- [x] New feature (change which adds functionality).
----------
## What's Changed:
- Added 2 `adopt_harvest` and `leave_harvest` views with their corresponding urls.
-  `adopt` checks if harvest has no pickleader, then it changes its status to `Adopted` and adds the user as a pickleader.
- `leave` does the vice-versa: checks if the user is the pickleader and so on...
- Added  a button beside the pickleader name to adopt/leave the harvest, doesn't get displayed if the harvest has another pickleader than the user.
- Added `harvest-status-change` view & url, also added `status_options` to the harvest detail context.
- Added a dropdown to change harvest status that appears if the user is the pickleader and if the harvest isn't orphaned.

--------
## Screenshots:

1.
![image](https://user-images.githubusercontent.com/52796958/179808130-40981d9e-4cce-4634-8f82-9e37d37df375.png)

2. trying to adopt harvest when there's already a pickleader
![image](https://user-images.githubusercontent.com/52796958/179808501-df92cf72-8c4c-42e6-876f-72a9016138cb.png)

3.
![image](https://user-images.githubusercontent.com/52796958/179807999-b61903f2-1108-48c6-a04c-862668196fbb.png)

4. Harvest status dropdown
![image](https://user-images.githubusercontent.com/52796958/179996937-1d35e987-5019-402f-b7dc-aa70e4dc4725.png)


--------
## URLs:
- http://127.0.0.1:8000/harvest/id/
- http://127.0.0.1:8000/harvest/adopt/id/
- http://127.0.0.1:8000/harvest/leave/id/
- http://127.0.0.1:8000/harvest/status-change/id/?status=

-------
## Checklist:
- [x] My code follows the [style guidelines](https://github.com/LesFruitsDefendus/saskatoon-ng/blob/develop/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.

mina-andajos commented:
> Hi @mina-andajos, 
> 
> Thanks for your work on this.
> 
> Is the development server up to date with this branch to try it out ? Can you create me an account on the development server and send me my credentials by email?
> 
> Thanks a lot,

hello tristan this branch is based off the master branch not develop, i dont think it will work with develop right off the bat also. i will create a super user for you though.

2ynn commented:
@mina-andajos since there haven't been any new migrations between master and develop, nothing prevents you from checking out this branch on the development server and reloading uwsgi to test it out

mina-andajos commented:
> @mina-andajos since there haven't been any new migrations between master and develop, nothing prevents you from checking out this branch on the development server and reloading uwsgi to test it out

hi @2ynn,
When I try to rebase a branch that's based off master, onto develop: for example `git rebase --onto=develop Feature-199-add-harvest-status-button-and-dropdown`, I get a lot of conflicts as the master branch still uses extra context and the old serializer while develop is using only the new harvest serializer. Which causes a lot of merge conflicts on my end. Am I doing it wrong?

2ynn commented on saskatoon/harvest/views.py:
```diff
@@ -277,3 +276,69 @@ def harvest_yield_create(request):
+    Used in a button at harvest-detail/status template
+    """
+    harvest = get_object_or_404(Harvest, id=id)
+    current_user = request.user
```
Here we should make sure that user belongs to either the 'pickleader' or 'core' role groups, e.g.:
```
if request.user.groups.filter(name__in=["pickleader", "core"]).exists():
```

2ynn commented on saskatoon/harvest/views.py:
```diff
@@ -277,3 +276,69 @@ def harvest_yield_create(request):
+    harvest = get_object_or_404(Harvest, id=id)
+    current_user = request.user
+
+    if harvest.pick_leader == None:
```
I believe PEP8 discourages ` == None`,  you could use `if harvest.pick_leader is None:` instead (or even `if not harvest.pick_leader:` in this case).

mina-andajos replied:
Thanks for the headsup.
I think `if not harvest.pick_leader:` is also [not a good practice](https://peps.python.org/pep-0008/#programming-recommendations)

mina-andajos replied:
Thanks for the headsup.
I think if not harvest.pick_leader: is also [not a good practice](https://peps.python.org/pep-0008/#programming-recommendations)

2ynn commented on saskatoon/harvest/views.py:
```diff
@@ -277,3 +276,69 @@ def harvest_yield_create(request):
+    Used in a button at harvest-detail/status template
+    """
+    harvest = get_object_or_404(Harvest, id=id)
+    current_user = request.user
+
+    if harvest.pick_leader == current_user:
```
I'd simply merge these two lines into:
```
if harvest.pick_leader == request.user:
```

2ynn commented on saskatoon/harvest/views.py:
```diff
@@ -277,3 +276,69 @@ def harvest_yield_create(request):
+        harvest.pick_leader = None
+        harvest.status = 'Orphan'
+        harvest.save()
+        messages.success(request, _("You Left the Harvest!"))
```
I'd suggest we don't capitalize "Left", e.g. : "You successfully left this harvest!", no big deal.

mina-andajos replied:
I have changed the messages 👍🏿

2ynn commented on saskatoon/sitebase/templates/app/detail_views/harvest/status.html:
```diff
@@ -47,14 +82,46 @@ <h2><span class="counter">{{ requests.count }}</span>
+                                    title="Leave it!"
+                                    class="btn"
+                                    >
+                                        <strong>{% trans "Leave it!" %}</strong>
```
I would personally use the same formatting as the Edit harvest button (currently not bold)

2ynn commented on saskatoon/sitebase/templates/app/detail_views/harvest/status.html:
```diff
@@ -10,11 +10,46 @@
+                                <span class="caret"></span>
+                            </h2>
+                            </button>
+                            <ul
+                            class="dropdown-menu"
+                            role="menu"
+                            aria-labelledby="menu1"
+                            >
+                                {% for status_option in status_options %}
+                                <li role="presentation">
+                                    <a
+                                    href="{% url 'harvest-status-change' harvest.id %}?status={{ status_option }}"
+                                    role="menuitem"
+                                    tabindex="-1"
+                                    >
+                                    {{ status_option }}
+                                </a>
+                                </li>
```
I get a bottom padding that is a bit too large on Firefox
![screenshot_2022-07-21_17:40:15](https://user-images.githubusercontent.com/41802320/180319116-8513deb0-cf79-4b45-aceb-d0d4fb9fd048.png)

Also would it be easy to change the color of the dropdown background based on the status to match the colors used in the calendar? I think it could be really useful.

mina-andajos replied:
I was thinking about that too. Great idea!

mina-andajos replied:
@2ynn What do you think about this?

![image](https://user-images.githubusercontent.com/52796958/180648942-9e3a5bd1-4bfb-4d4d-8a50-6caf9e4d02d6.png)
![image](https://user-images.githubusercontent.com/52796958/180648961-ef50f74c-bc4e-4496-bcea-861b7ab18a5f.png)
![image](https://user-images.githubusercontent.com/52796958/180648983-456cb9ee-7e1f-411e-a34f-e61b1275aa53.png)

**The rest of the statuses**:
![image](https://user-images.githubusercontent.com/52796958/180648995-d3f84224-f60e-4999-9c1a-e1e1ad1af32e.png)

mina-andajos commented:
@2ynn hi
could you take a final look and merge?

2ynn reviewed (approved):
Just a little comment, I find the `user_is_core` variable name slightly confusing since we are checking if user is "at least" a pickleader, and doesn't have to be a core member for the boolean to be true. Other than that looks really great! thanks a lot

mina-andajos commented:
> 



> Just a little comment, I find the `user_is_core` variable name slightly confusing since we are checking if user is "at least" a pickleader, and doesn't have to be a core member for the boolean to be true. Other than that looks really great! thanks a lot

I changed it to `user_is_core_or_pick` and added a comment to make it clearer.

2ynn commented:
FYI just updated prod server with these changes
